### ISSUE TITLE
Training Script

### ISSUE THREAD

**Issue by adityac8** [bradezard131/wsod#1]
Hi @bradezard131 
Thank you for such amazing work.
As per the README, I made my environment and have the latest Detectron2 and all its dependencies.
I use the following training script 
```
python tools/train_net.py \
	--config-file configs/pcl_vgg_dilated.yaml
```
The code runs on a single gpu and gives the following results
```
[02/01 02:23:28 d2.evaluation.testing]: copypaste: Task: bbox
[02/01 02:23:28 d2.evaluation.testing]: copypaste: AP,AP50,AP75
[02/01 02:23:28 d2.evaluation.testing]: copypaste: 17.3432,40.9063,12.2699
```

However, the results reported by PCL as per the README is `44.21`.
Could you please tell me any what training script or any modifications(such as _images per batch_ or _number of gpus_ ) you use to replicate the results.

**Comment by bradezard131** [comment]
Hi,

Looks like I've had a few hitches in cleaning and moving this to a public repo (this is a mirror of a private repo where I do my research). I seem to have achieved back in range by increasing ITER_SIZE to 4 (like in Peng Tang's repo) and doubling the steps (so the scheduler steps becomes 70000 rather than 35000, and max_iter is 100000 not 50000). This is pretty close then to the hyperparams in Peng's repo (since their config file gets dynamically edited by the program when it runs). I'm still re-running it again to confirm (I've misplaced a few things it seems and am trying to reproduce all my old results again), but if you don't want to wait that is my recommendation

**Comment by adityac8** [comment]
Hi,
Thanks for the settings.
I run the code with the modifications and get the following results:
```
[02/10 07:48:57 d2.evaluation.testing]: copypaste: Task: bbox
[02/10 07:48:57 d2.evaluation.testing]: copypaste: AP,AP50,AP75
[02/10 07:48:57 d2.evaluation.testing]: copypaste: 19.1267,42.9265,14.1507
```
The results are for single-gpu training. I would try running it on multiple-gpus and see if it further improves.

**Comment by adityac8** [comment]
I make the modifications as told by detectron2 docs for multi-GPU but end up getting mAP around 40.
I use the following settings,
```
SOLVER:
  IMS_PER_BATCH: 8 # 1 multiply by 8
  ITER_SIZE: 4
  BASE_LR: 0.004 # 5e-4 multiply by 8
  WEIGHT_DECAY: 0.004 # 5e-4 multiply by 8
  BIAS_LR_FACTOR: 2.0
  WEIGHT_DECAY_BIAS: 0.0
  REFINEMENT_LR_FACTOR: 1.0
  MOMENTUM: 0.9
  LR_SCHEDULER_NAME: "WarmupMultiStepLR"
  STEPS: (8750,) #70000 divide by 8
  GAMMA: 0.1
  WARMUP_FACTOR: 1.
  WARMUP_ITERS: 0
  MAX_ITER: 12500 #100000 divide by 8
  CHECKPOINT_PERIOD: 625 #5000 divide by 8
```
I manage to get
```
[02/14 15:54:33 d2.evaluation.testing]: copypaste: Task: bbox
[02/14 15:54:33 d2.evaluation.testing]: copypaste: AP,AP50,AP75
[02/14 15:54:33 d2.evaluation.testing]: copypaste: 17.9214,40.0014,13.5763
```

However, changing the iters to default (same as voc as given by detectron2) for 8 gpus,
```
SOLVER:
  IMS_PER_BATCH: 8
  ITER_SIZE: 4
  BASE_LR: 0.004 # 5e-4 multiply by 8
  WEIGHT_DECAY: 0.004 # 5e-4 multiply by 8
  BIAS_LR_FACTOR: 2.0
  WEIGHT_DECAY_BIAS: 0.0
  REFINEMENT_LR_FACTOR: 1.0
  MOMENTUM: 0.9
  LR_SCHEDULER_NAME: "WarmupMultiStepLR"
  STEPS: (12000,16000)
  GAMMA: 0.1
  WARMUP_FACTOR: 1.
  WARMUP_ITERS: 0
  MAX_ITER: 18000
  CHECKPOINT_PERIOD: 1000
```
I manage to get
```
[02/14 14:54:26 d2.evaluation.testing]: copypaste: Task: bbox
[02/14 14:54:26 d2.evaluation.testing]: copypaste: AP,AP50,AP75
[02/14 14:54:26 d2.evaluation.testing]: copypaste: 19.2759,41.2105,15.5178
```
Could you please provide the settings for 8-gpus in order to replicate 44.21
Thanks

**Comment by bradezard131** [comment]
As stated above I get my 44 mAP with single GPU training and the listed modifications. Both my results and Tang et. al. are produced with an effective batch size of 4. To do 8 gpu training, you'd need to have a batch size of at least 8, so you'll have to work out the other hyperparameters yourself. I don't have access to 8 gpus to do this

**Comment by adityac8** [comment]
Hi,
Could you please check that with 1-gpu settings, if you are able to replicate 44.
I just saw, the [losses.py](https://github.com/bradezard131/wsod/blob/master/wsod/models/losses.py) file has num_clusters as 5. I think changing that to 3 would give the results.
Thanks

**Comment by bradezard131** [comment]
I am able to replicate the settings with the previously described changes. The things in losses.py are directly from the original author's repo

**Comment by adityac8** [comment]
Ok,
I'll try changing some dependency versions to see if I get the numbers.
Do you use the proposals as given by https://github.com/ppengtang/pcl.pytorch
Also, a few scores in each `<classname>.txt` in `tmp/dets` seem to be higher than 1. Could you please tell me why is it happening?
Thanks

**Comment by bradezard131** [comment]
Scores are sums of each of the 5 scales and flips, rather than the mean. It makes no real difference whether you use mean or sum

**Comment by tianyuecao** [comment]
> Ok,
> I'll try changing some dependency versions to see if I get the numbers.
> Do you use the proposals as given by https://github.com/ppengtang/pcl.pytorch
> Also, a few scores in each `<classname>.txt` in `tmp/dets` seem to be higher than 1. Could you please tell me why is it happening?
> Thanks

Hi, I'm trying to run this code, but I can't find the proposal file. Which proposal file did you use? I tired the proposals given by https://github.com/ppengtang/pcl.pytorch but I got index errors.
Thanks!

**Comment by adityac8** [comment]
Hi
I used the same proposal file as given in the `pcl.pytorch` code.
It works for me.
Thanks

**Comment by bradezard131** [comment]
The same pregenerated proposals that everyone uses. You may need to convert them into detectron2 formatting which I'm pretty sure is documented somewhere in there.

**Comment by xiaaoo** [comment]
> > Ok,
> > I'll try changing some dependency versions to see if I get the numbers.
> > Do you use the proposals as given by https://github.com/ppengtang/pcl.pytorch
> > Also, a few scores in each `<classname>.txt` in `tmp/dets` seem to be higher than 1. Could you please tell me why is it happening?
> > Thanks
> 
> Hi, I'm trying to run this code, but I can't find the proposal file. Which proposal file did you use? I tired the proposals given by https://github.com/ppengtang/pcl.pytorch but I got index errors.
> Thanks!

I got index error too using the proposals given by `pcl.pytorch`. Did you manage to fix it?
